// @makerdao/dai/contracts/abi/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.general = exports.exchangesV1 = exports.dappHub = exports.daiV1 = undefined;

var _v = require('./erc20Token/v1.json');

var _v2 = _interopRequireDefault(_v);

var _v3 = require('./dsEthToken/v1.json');

var _v4 = _interopRequireDefault(_v3);

var _DSValue = require('./dai/v1/DSValue.json');

var _DSValue2 = _interopRequireDefault(_DSValue);

var _DSGuard = require('./dai/v1/DSGuard.json');

var _DSGuard2 = _interopRequireDefault(_DSGuard);

var _DSChief = require('./dai/v1/DSChief.json');

var _DSChief2 = _interopRequireDefault(_DSChief);

var _DSSpell = require('./dai/v1/DSSpell.json');

var _DSSpell2 = _interopRequireDefault(_DSSpell);

var _DSSpellBook = require('./dai/v1/DSSpellBook.json');

var _DSSpellBook2 = _interopRequireDefault(_DSSpellBook);

var _v5 = require('./makerOtc/v1.json');

var _v6 = _interopRequireDefault(_v5);

var _v7 = require('./zeroExExchange/v1.json');

var _v8 = _interopRequireDefault(_v7);

var _SaiTop = require('./dai/v1/SaiTop.json');

var _SaiTop2 = _interopRequireDefault(_SaiTop);

var _SaiTub = require('./dai/v1/SaiTub.json');

var _SaiTub2 = _interopRequireDefault(_SaiTub);

var _SaiTap = require('./dai/v1/SaiTap.json');

var _SaiTap2 = _interopRequireDefault(_SaiTap);

var _SaiVox = require('./dai/v1/SaiVox.json');

var _SaiVox2 = _interopRequireDefault(_SaiVox);

var _SaiMom = require('./dai/v1/SaiMom.json');

var _SaiMom2 = _interopRequireDefault(_SaiMom);

var _GemPit = require('./dai/v1/GemPit.json');

var _GemPit2 = _interopRequireDefault(_GemPit);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var daiV1 = {
  saiTop: _SaiTop2.default,
  tub: _SaiTub2.default,
  tap: _SaiTap2.default,
  vox: _SaiVox2.default,
  mom: _SaiMom2.default,
  pit: _GemPit2.default
};

var dappHub = {
  dsValue: _DSValue2.default,
  dsEthToken: _v4.default,
  dsGuard: _DSGuard2.default,
  dsChief: _DSChief2.default,
  dsSpell: _DSSpell2.default,
  dsSpellBook: _DSSpellBook2.default
};

var exchangesV1 = {
  zeroExExchange: _v8.default,
  makerOtc: _v6.default
};

var general = {
  erc20: _v2.default
};

exports.daiV1 = daiV1;
exports.dappHub = dappHub;
exports.exchangesV1 = exchangesV1;
exports.general = general;
