// @makerdao/dai/src/eth/PriceService.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _PrivateService2 = require('../core/PrivateService');

var _PrivateService3 = _interopRequireDefault(_PrivateService2);

var _contracts = require('../../contracts/contracts');

var _contracts2 = _interopRequireDefault(_contracts);

var _constants = require('../utils/constants');

var _bignumber = require('bignumber.js');

var _bignumber2 = _interopRequireDefault(_bignumber);

var _ethers = require('ethers');

var _Currency = require('./Currency');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PriceService = function (_PrivateService) {
  (0, _inherits3.default)(PriceService, _PrivateService);

  /**
   * @param {string} name
   */

  function PriceService() {
    var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'price';
    (0, _classCallCheck3.default)(this, PriceService);
    return (0, _possibleConstructorReturn3.default)(this, (PriceService.__proto__ || (0, _getPrototypeOf2.default)(PriceService)).call(this, name, ['token', 'smartContract', 'event']));
  }

  (0, _createClass3.default)(PriceService, [{
    key: 'initialize',
    value: function initialize() {
      var _this2 = this;

      this.get('event').registerPollEvents({
        'price/ETH_USD': {
          price: function price() {
            return _this2.getEthPrice();
          }
        },
        'price/MKR_USD': {
          price: function price() {
            return _this2.getMkrPrice();
          }
        },
        'price/WETH_PETH': {
          ratio: function ratio() {
            return _this2.getWethToPethRatio();
          }
        }
      });
    }
  }, {
    key: '_getContract',
    value: function _getContract(contract) {
      return this.get('smartContract').getContractByName(contract);
    }
  }, {
    key: '_valueForContract',
    value: function _valueForContract(value, unit) {
      var bn = (0, _Currency.getCurrency)(value, unit).toEthersBigNumber('wei');
      return _ethers.utils.hexlify(_ethers.utils.padZeros(bn, 32));
    }
  }, {
    key: 'getWethToPethRatio',
    value: function getWethToPethRatio() {
      return this._getContract(_contracts2.default.SAI_TUB).per().then(function (bn) {
        return new _bignumber2.default(bn.toString()).dividedBy(_constants.RAY).toNumber();
      });
    }
  }, {
    key: 'getEthPrice',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee() {
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.t0 = _Currency.USD_ETH;
                _context.next = 3;
                return this._getContract(_contracts2.default.SAI_PIP).read();

              case 3:
                _context.t1 = _context.sent;
                return _context.abrupt('return', _context.t0.wei.call(_context.t0, _context.t1));

              case 5:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function getEthPrice() {
        return _ref.apply(this, arguments);
      }

      return getEthPrice;
    }()
  }, {
    key: 'getPethPrice',
    value: function () {
      var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2() {
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.t0 = _Currency.USD_PETH;
                _context2.next = 3;
                return this._getContract(_contracts2.default.SAI_TUB).tag();

              case 3:
                _context2.t1 = _context2.sent;
                return _context2.abrupt('return', _context2.t0.ray.call(_context2.t0, _context2.t1));

              case 5:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getPethPrice() {
        return _ref2.apply(this, arguments);
      }

      return getPethPrice;
    }()
  }, {
    key: 'getMkrPrice',
    value: function () {
      var _ref3 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3() {
        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.t0 = _Currency.USD_MKR;
                _context3.next = 3;
                return this._getContract(_contracts2.default.SAI_PEP).peek();

              case 3:
                _context3.t1 = _context3.sent[0];
                return _context3.abrupt('return', _context3.t0.wei.call(_context3.t0, _context3.t1));

              case 5:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getMkrPrice() {
        return _ref3.apply(this, arguments);
      }

      return getMkrPrice;
    }()
  }, {
    key: 'setEthPrice',
    value: function setEthPrice(newPrice) {
      var unit = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _Currency.ETH;

      var value = this._valueForContract(newPrice, unit);
      return this._getContract(_contracts2.default.SAI_PIP).poke(value);
    }
  }, {
    key: 'setMkrPrice',
    value: function setMkrPrice(newPrice) {
      var unit = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _Currency.MKR;

      var value = this._valueForContract(newPrice, unit);
      return this._getContract(_contracts2.default.SAI_PEP).poke(value);
    }
  }]);
  return PriceService;
}(_PrivateService3.default);

exports.default = PriceService;
